File: src/main/java/pl/dmichalski/algorithms/_9_sorting/bubble/BubbleSortService.java
package pl.dmichalski.algorithms._9_sorting.bubble;

import java.util.ArrayList;
import java.util.List;

class BubbleSortService {

    /**
     * O(n2) complexity
     */
    List<Integer> sort(List<Integer> values) {
        if (values == null) {
            return null;
        }

        List<Integer> copyOfValues = new ArrayList<>(values);
        int n = copyOfValues.size();

        for (int i = 0; i < n - 1; i++) {
            for (int j = 0; j < n - i - 1; j++) {
                if (copyOfValues.get(j) > copyOfValues.get(j + 1)) {
                    int tmp = copyOfValues.get(j);
                    copyOfValues.set(j, copyOfValues.get(j + 1));
                    copyOfValues.set(j + 1, tmp);
                }
            }
        }

        return copyOfValues;
    }

}

File: src/main/kotlin/pl/dmichalski/algorithms/_9_sorting/bubble/BubbleSortService.kt
package pl.dmichalski.algorithms._9_sorting.bubble

import pl.dmichalski.algorithms._9_sorting.SortService

/**
 * Bubble sort algorithm.
 *
 * O(n2) time complexity
 * O(1) space complexity
 */
internal class BubbleSortService : SortService {

    override fun sort(values: IntArray): IntArray {
        val valuesCopy = values.copyOf()
        val n = valuesCopy.size
        for (i in 0 until n - 1) {
            for (j in 0 until n - i - 1) {
                if (valuesCopy[j] > valuesCopy[j + 1]) {
                    val tmp = valuesCopy[j]
                    valuesCopy[j] = valuesCopy[j + 1]
                    valuesCopy[j + 1] = tmp
                }
            }
        }

        return valuesCopy
    }

}
